### DISCUSSION TITLE
Force keyword arguments

### DISCUSSION THREAD

**Discussion by xLaszlo** [astral-sh/ruff#8137] [Ideas]
This might be addressed elsewhere, but how do I force keyword arguments or reformat the code to add them automatically? 

This is what I found for Flake:

https://pypi.org/project/flake8-force-keyword-arguments/

I'd like to see the same behaviour, maybe with automatically adding them at fixing.

**Comment by charliermarsh** [comment]
I believe this plugin only checks for functions defined in the same module as the usage. Is that correct? (I tested it too, and it seems to be the case.) We could support it, but it seems quite limited.

**Comment by cleder** [comment]
It does check for functions outside the same module, e.g:
```
pygeoif/factories.py:46:27: FKA100 re.compile's call uses 2 positional arguments, use keyword arguments.
pygeoif/functions.py:150:25: FKA100 zip_longest's call uses 2 positional arguments, use keyword arguments.
pygeoif/functions.py:158:20: FKA100 math.isclose's call uses 2 positional arguments, use keyword arguments.
...
```

An autofix would be nice, even if that would be limited to the same module as the usage

**Comment by xLaszlo** [comment]
Yes, I think it checks other modules as well. (Thanks for dealing with this, especially this fast!)

I used the following command to run it for a project and then increased the argument and checked the relevant mistakes. 

`flake8 NewVersion/ --select FKA1 --kwargs-max-positional-arguments 0` 

After that, I ditched it because I didn't want to flake in the project just. 

I am not sure what is the general policy on this. It probably would be great to enforce it or project files, but not for built-in or common commands. There are a lot of those errors that @cleder has shown.

**Reply by cleder** [reply]
- `re.compile`  can be `re.compile(pattern= ..., flags=...)` :heavy_check_mark: 
- `itertools.zip_longest` accepts positional arguments only for the values to be zipped, and forces keyword arguments for `fillvalue` :heavy_multiplication_x: 
- `math.isclose` would be `math.isclose(a=.., b=...)` which does not improve readability :heavy_multiplication_x: 

I do not think it is worth the effort at this time

**Comment by rrhodes** [comment]
This sounds like a duplicate of #3269, unless I'm mistaken?

**Comment by xLaszlo** [comment]
Not sure if this is relevant, but our standard policy for selectively enforcing keywords is `def f(*,...)`. The `*` captures all non-keyword arguments and forces the explicit writing of all subsequent ones. Then we enforce the presence of the `*` in selected modules (in codereview).

**Reply by daniel-flexa** [reply]
Adding `*` works in your own codebase, but not for calls to third party functions.

**Reply by xLaszlo** [reply]
That is very true.

**Comment by rrhodes** [comment]
Would PRs for this issue be welcome from community contributions? I recognise it's neither labelled `good first issue` nor `help wanted` as per [the Contributing guide](https://docs.astral.sh/ruff/contributing/) and it sounds like this issue would be non-trivial to resolve, so contributions may be valuable only from experienced community members?

**Reply by ntBre** [reply]
https://github.com/astral-sh/ruff/issues/3269 is still labeled `needs-decision`, so I think we're not fully sold on adding these rules at the moment. We're working on multi-file analysis in ty, so it might make sense to revisit this once we have access to that infrastructure in Ruff.
